src/axom/inlet/tests/inlet_object.cpp

```diff
@@ -428,6 +426,114 @@ TEST(inlet_object, implicit_conversion_primitives)
+
+  inlet.addIntDict("foo", "foo's description");
+  std::unordered_map<std::string, int> dict = inlet["foo"];
+  std::unordered_map<std::string, int> correct = {{"key1", 4},
+                                                  {"key3", 6},
+                                                  {"key2", 10}};
+  EXPECT_EQ(dict, correct);
+}
+
+TEST(inlet_dict, simple_dict_of_struct_by_value)
+{
+  std::string testString =
+    "foo = { [\"key1\"] = { bar = true; baz = false}, "
+    "        [\"key2\"] = { bar = false; baz = true} }";
+  DataStore ds;
+  auto inlet = createBasicInlet(&ds, testString);
+
+  auto& dict_table = inlet.addGenericDict("foo");
+
+  dict_table.addBool("bar", "bar's description");
+  dict_table.addBool("baz", "baz's description");
+  std::unordered_map<std::string, Foo> expected_foos = {{"key1", {true, false}},
+                                                        {"key2", {false, true}}};
+  std::unordered_map<std::string, Foo> foos;
+  foos = inlet["foo"].get<std::unordered_map<std::string, Foo>>();
+  EXPECT_EQ(foos, expected_foos);
+}
+
+struct FooWithDict
+{
+  std::unordered_map<std::string, int> arr;
+  bool operator==(const FooWithDict& other) const { return arr == other.arr; }
+};
+
+template <>
+struct FromInlet<FooWithDict>
+{
+  FooWithDict operator()(const axom::inlet::Table& base)
+  {
+    FooWithDict f = {base["arr"]};
+    return f;
+  }
+};
+
+TEST(inlet_dict, dict_of_struct_containing_dict)
+{
+  std::string testString =
+    "foo = { [\"key3\"] = { arr = { [\"key1\"] = 3 }; }, "
+    "        [\"key4\"] = { arr = { [\"key2\"] = 2 }; } }";
+  DataStore ds;
+  auto inlet = createBasicInlet(&ds, testString);
+
+  auto& dict_table = inlet.addGenericDict("foo");
+
+  dict_table.addIntDict("arr", "arr's description");
+  std::unordered_map<std::string, FooWithDict> expected_foos = {
+    {"key3", {{{"key1", 3}}}},
+    {"key4", {{{"key2", 2}}}}};
+  std::unordered_map<std::string, FooWithDict> foos_with_dict;
+  foos_with_dict =
+    inlet["foo"].get<std::unordered_map<std::string, FooWithDict>>();
+  EXPECT_EQ(foos_with_dict, expected_foos);
+}
+
+TEST(inlet_dict, dict_of_struct_containing_array)
+{
+  std::string testString =
+    "foo = { [\"key3\"] = { arr = { [1] = 3 }; }, "
+    "        [\"key4\"] = { arr = { [6] = 2 }; } }";
+  DataStore ds;
+  auto inlet = createBasicInlet(&ds, testString);
+
+  auto& dict_table = inlet.addGenericDict("foo");
+
+  dict_table.addIntArray("arr", "arr's description");
+  std::unordered_map<std::string, FooWithArray> expected_foos = {
+    {"key3", {{{1, 3}}}},
+    {"key4", {{{6, 2}}}}};
+  std::unordered_map<std::string, FooWithArray> foos_with_array;
+  foos_with_array =
+    inlet["foo"].get<std::unordered_map<std::string, FooWithArray>>();
+  EXPECT_EQ(foos_with_array, expected_foos);
+}
+
+TEST(inlet_dict, array_of_struct_containing_dict)
+{
+  std::string testString =
+    "foo = { [7] = { arr = { [\"key1\"] = 3 }; }, "
```

Comment:
What would happen if the key has a slash ("/") in it, 
e.g.
```cpp
std::string testString =
    "foo = { [7] = { arr = { ['key1'] = 3 }; },
             [4] = {arr = { ['key1/subkey1'] = 3 }; }, "
```
Is this worth testing?

Reply:
More generally,  are there restrictions on the key strings?

Reply:
This was a good idea - the current behavior is to add a subtable (which is in general inaccessible) to the array table.  I've added an error when the key contains a forward slash.

I tried a couple other potentially problematic keys and discovered that (unsurprisingly) the empty string is not a valid key, so I've added an error for that as well.

If a key that contains a slash is valid, it might be worth looking at user-specified scope delimiters (instead of hardcoded forward-slash) in a future PR.  I'm not sure whether this is an inlet-only change or if the forward-slash is hardcoded in Sidre as well.

Reply:
Re: sidre -- I think we added some knobs to theoretically allow for it to be user-changeable, but in practice that has not been explored, so it would likely need to fine-tuning / debugging to get that working.